lib/ipam/ipam_test.go

```diff
@@ -678,6 +678,88 @@ var _ = testutils.E2eDatastoreDescribe("IPAM tests", testutils.DatastoreEtcdV3,
 			node2IP := v4[0].IP
 			Expect(pool1.IPNet.Contains(node2IP)).To(BeTrue(), fmt.Sprintf("%s not in pool %s", node2IP, pool1))
 		})
+
+		// Allocates IPs from a pool that has a matching node selector,
+		// deallocates them all, deselects the pool from the node,
+		// tests that the block affinity should be released.
+		It("should release affinity and block when pool is empty and node selector is deselected", func() {
+			host := "host"
+			pool1 := cnet.MustParseNetwork("10.0.0.0/24")
+
+			bc.Clean()
+			deleteAllPools()
+
+			applyNode(bc, host, map[string]string{"foo": "bar"})
+			applyPool(pool1.String(), true, `foo == "bar"`)
+
+			// Assign three addresses to the node.
+			v4nets, _, err := ic.AutoAssign(context.Background(), AutoAssignArgs{
+				Num4:     3,
+				Num6:     0,
+				Hostname: host,
+			})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(v4nets)).To(Equal(3))
+
+			// Should have one affine block to this host.
+			blocks := getAffineBlocks(bc, host)
+			Expect(len(blocks)).To(Equal(1))
+
+			// Expect all the IPs to be from pool1.
+			var v4IPs []cnet.IP
+			for _, a := range v4nets {
+				Expect(pool1.IPNet.Contains(a.IP)).To(BeTrue(), fmt.Sprintf("%s not in pool %s", a.IP, pool1))
+				v4IPs = append(v4IPs, cnet.IP{a.IP})
+			}
+
+			// Release all IPs.
+			unallocated, err := ic.ReleaseIPs(context.Background(), v4IPs)
+			Expect(len(unallocated)).To(Equal(0))
+			Expect(err).NotTo(HaveOccurred())
+
+			// Change the selector for the IP pool so that it no longer matches node1.
+			applyPool(pool1.String(), true, `foo != "bar"`)
+
+			// The block should still have an affinity to this host.
+			Expect(len(getAffineBlocks(bc, host))).To(Equal(1))
+
+			// The block should still have affinity.
+			opts := model.BlockListOptions{IPVersion: 4}
```

Comment:
This bit of code actually isn't doing what the comment says - it's querying blocks rather than affinities, and asserting that a single block exists but not doing any assertion on the affinity.

Reply:
addressed in a808da3